runtime/commonMain/src/kotlinx/serialization/internal/CollectionSerializers.kt

```diff
@@ -140,6 +140,45 @@ sealed class MapLikeSerializer<TKey, TVal, TCollection, TBuilder: MutableMap<TKe
     }
 }
 
+abstract class PrimitiveArraySerializer<TElement, TCollection, TBuilder : PrimitiveArraySerializer<TElement, TCollection, TBuilder>.Builder>
+internal constructor(
+    primitiveSerializer: KSerializer<TElement>,
+    primitiveDescriptor: PrimitiveDescriptor
+) : ListLikeSerializer<TElement, TCollection, TBuilder>(primitiveSerializer) {
+    final override val descriptor = PrimitiveArrayDescriptor(primitiveDescriptor)
+
+    abstract inner class Builder internal constructor() {
+        abstract val pos: Int
+        abstract fun ensureCapacity(requiredCapacity: Int = pos + 1)
+        abstract fun build(): TCollection
+    }
+
+    final override fun TBuilder.builderSize() = pos
+    final override fun TBuilder.toResult(): TCollection = build()
+    final override fun TBuilder.checkCapacity(size: Int) = ensureCapacity(size)
+
+    // these methods lead to boxing
+    final override fun TCollection.objIterator(): Iterator<TElement> = throw AssertionError("Must not be used")
```

Comment:
Why not? Document this aspect properly.
Please use `error()` to throw ISE, AE is used when internal invariant is violated, not during API misuse